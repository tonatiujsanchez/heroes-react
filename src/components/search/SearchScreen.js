import { useEffect, useState } from "react";
import { useNavigate, useLocation } from "react-router-dom";
import queryString from 'query-string'

import { useForm } from "../../hooks/useForm";

import { HeroCard } from "../hero/HeroCard";
import { getHeroesByName } from "../../selectors/getHeroesByName";


const SearchScreen = () => {

    const navigate = useNavigate()
    const location = useLocation()

    const { q = '' } = queryString.parse(location.search)
    
    const [heroesSearch, setHeroesSearch] = useState([])
    


    const [ { searchText }, handleInputChange ] = useForm({
        searchText: q
    })
    
    useEffect(()=>{
        const heroesFilter = getHeroesByName( searchText.trim() )
        setHeroesSearch(heroesFilter)
    },[])
    


    const handleSubmit = ( e )=>{
        e.preventDefault()
        
        if( searchText.trim() === '' ){
            console.log('Se necesita una temino para realizar la busqueda');
            return
        }
        if( searchText.trim() === q ){
            return
        }        

        navigate(`?q=${ searchText.trim() }`)

        const heroesFilter = getHeroesByName( searchText.trim() )
        setHeroesSearch( heroesFilter );
    }

    return (
        <>
            <h1>Buscar Heroes</h1>
            <hr />
            <div className="row mb-5">
                <div className="col-5">
                    <form onSubmit={ handleSubmit }>
                        <input
                            type="text"
                            placeholder="Batman, Wonder Woman, Spider Man, Captain America, etc."
                            className="form-control mb-3"
                            name="searchText"
                            value={ searchText }
                            onChange={ handleInputChange }
                            autoComplete="off" />
                        <div className="d-grid gap-2">
                            <button 
                                type="submit" 
                                className="btn btn-success w-block">
                                    Buscar Heroe
                            </button>
                        </div>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Resultados</h4>
                    <hr/>
                    { heroesSearch.length > 0 && (
                        
                        heroesSearch.map( hero => (
                            <div key={ hero.id } className="mb-3">
                                <HeroCard hero={ hero } />
                            </div>
                            ))
                      )
                    }
                    { q === '' 
                        ? ( <div className="alert alert-info text-center">Busca un h√©roe</div> )
                        : q !== '' && (heroesSearch.length === 0) && (
                            <div className="alert alert-info text-center">
                                No se encontraron resultados de <span className="fw-bold">{ q }</span>
                            </div>
                        )  
                    }
  
                </div>
            </div>
        </>
    )
}

export default SearchScreen