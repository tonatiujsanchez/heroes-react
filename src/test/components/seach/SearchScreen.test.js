import { mount } from "enzyme"
import { MemoryRouter } from "react-router-dom"
import SearchScreen from "../../../components/search/SearchScreen"


const mockNavigate = jest.fn()

jest.mock('react-router-dom', ()=>({
    ...jest.requireActual('react-router-dom'),
    useNavigate: ()=> mockNavigate,
}))


describe('Pruebas a <SearchScreen />', () => {

    test('Debe mostrarse correctamente con valores por defecto', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('.alert-info').text().trim()).toBe('Busca un hÃ©roe')

    })


    test('Debe de mostrar a Batman y el input cib ek valir del query String', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>
        )

        expect( wrapper ).toMatchSnapshot()
        expect( wrapper.find('input').prop('value') ).toBe('batman')


    })


    test('Debe de mostrar en mensaje de Sin resultados', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=123456']}>
                <SearchScreen />
            </MemoryRouter>
        )

        expect( wrapper ).toMatchSnapshot()
        
        const msjSinResultados = wrapper.find('.alert-info').text().trim()
        expect( msjSinResultados ).toBe('No se encontraron resultados de 123456')

    })

    test('Debe de llamar el navigate a la nueva pantalla', () => { 

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        )

        wrapper.find('input').simulate('change', { 
            target:{ name: 'searchText', value: 'batman' } 
        })

        wrapper.find('form').prop('onSubmit')({
            preventDefault: ()=>{}
        })

        expect( mockNavigate ).toHaveBeenCalledWith('?q=batman')

     })


    
})