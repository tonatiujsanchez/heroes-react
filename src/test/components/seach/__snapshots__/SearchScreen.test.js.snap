// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas a <SearchScreen /> Debe de mostrar a Batman y el input cib ek valir del query String 1`] = `
Array [
  <h1>
    Buscar Heroes
  </h1>,
  <hr />,
  <div
    className="row mb-5"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control mb-3"
          name="searchText"
          onChange={[Function]}
          placeholder="Batman, Wonder Woman, Spider Man, Captain America, etc."
          type="text"
          value="batman"
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn btn-success w-block"
            type="submit"
          >
            Buscar Heroe
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="mb-3"
        key="dc-batman"
      >
        <div
          className="col"
        >
          <div
            className="card animate__animated animate__fadeIn"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Batman"
                  className="card-img"
                  src="/assets/img/dc-batman.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-title"
                  >
                    Batman
                  </h5>
                  <p
                    className="card-text"
                  >
                    Bruce Wayne
                  </p>
                  <p
                    className="card-text"
                  >
                    Detective Comics #27
                  </p>
                  <a
                    href="/hero/dc-batman"
                    onClick={[Function]}
                  >
                    Ver más
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas a <SearchScreen /> Debe de mostrar en mensaje de Sin resultados 1`] = `
Array [
  <h1>
    Buscar Heroes
  </h1>,
  <hr />,
  <div
    className="row mb-5"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control mb-3"
          name="searchText"
          onChange={[Function]}
          placeholder="Batman, Wonder Woman, Spider Man, Captain America, etc."
          type="text"
          value="123456"
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn btn-success w-block"
            type="submit"
          >
            Buscar Heroe
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info text-center"
      >
        No se encontraron resultados de 
        <span
          className="fw-bold"
        >
          123456
        </span>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas a <SearchScreen /> Debe mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Buscar Heroes
  </h1>,
  <hr />,
  <div
    className="row mb-5"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control mb-3"
          name="searchText"
          onChange={[Function]}
          placeholder="Batman, Wonder Woman, Spider Man, Captain America, etc."
          type="text"
          value=""
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn btn-success w-block"
            type="submit"
          >
            Buscar Heroe
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info text-center"
      >
        Busca un héroe
      </div>
    </div>
  </div>,
]
`;
